Known Customer Credential-Issuer (#1330)
TBD54566975/developer.tbd.website: site/docs/tbdex/issuer/kcc/kcc-issuer.mdx

For a comprehensive exploration of the practical applications and compliance considerations of KCCs, refer to our [KCC Compliance Guide](/) which provides detailed examples and insights into how KCCs align with certain regulatory requirements, offering valuable context for those overseeing the implementation and management of KCCs.

:::danger Important
This guide is intended for educational purposes only and does not constitute legal advice. Compliance programs may have varying requirements. Consult your legal and/or compliance advisors to ensure that the KCC is consistent with your legal and compliance obligations.
:::

In this guide, we'll cover:

- Performing Identity Verification (IDV) as a necessary step **before** issuing a Known Customer Credential (KCC), by implementing an API conformant with [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) and [OpenID for Verifiable Presentation's (OID4VP)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html).
- Designing a KCC, detailing the required properties.

> - Issuing a KCC, conformant with [OID4VCI](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html), following successful IDV.

Reviewer: To be consistent with the first bullet point:

```suggestion
- Issuing a KCC, conformant with [OpenID for Verifiable Credential Issuance (OID4VCI)](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html), following successful IDV.
```

:::info
For a detailed guide on managing Known Customer Credentials as a Wallet application, please refer to the [Known Customer Credential Guide for Wallet Apps](/)
:::

## Environment Setup

If you haven't already, please follow the [Credential Issuance Server Setup Guide](/) for detailed instructions on the dependencies and packages needed to set up your server.

## IDV Process

Identity Verification (IDV) is a critical component of typical Know Your Customer (KYC) requirements, where certain types of Personally Identifying Information (PII) are collected and verified from an individual. IDV is a crucial step before a Known Customer Credential (KCC) can be issued.

[**IDV Flow diagram**]

The IDV flow begins when the customer's agent (e.g. Wallet application) sends an HTTP request to the Issuer's [IDV service endpoint](https://hackmd.io/4cW3D9BVQGCEuf1CEg47jA?both#Issuers-Identity) specified in the Issuer's [Decentralized Identifier (DID) Document](/docs/glossary#did-document).

### Handling Incoming Requests

> 1. Implement an endpoint to handle incoming GET requests from the Wallet. When a request is received, you'll need to construct the [SIOPv2 Authorization Request](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-9).

Reviewer 2: this is the right link, just calling out it's important we reference the same draft across our docs

Reviewer: I had the same thought, and it's going to be difficult to keep these up to date. @decentralgabe I noticed OID4VCI and OID4VP urls are both 1.0 even tho they are marked as a draft, but then the SIOPv2 URL has "draft" within it. Do you know if there is a 1.0 equivalent for the SIOPv2 spec?

Else @EbonyLouis we're going to have to keep an eye on this and once new drafts/URLs are published we'll have to go back through everywhere and update

}
```

The decoded [JWT Claims](https://datatracker.ietf.org/doc/html/rfc7519#section-4) for the `id_token` and `vp_token` are as follows:

```javascript
// Wallet.js id_token decoded JWT claims

{
  "iss": "did:dht:customerDid",
  "sub": "did:dht:customerDid", // Customer's DID string
  "aud": "did:dht:issuerDid", // Issuer's DID string
  "nonce": "n-0S6_WzA2Mj", // Unique string from the Issuer's Auth request
  "exp": 1618884473, // Expiration time
  "iat": 1618880873 // issued at time
}
```

```javascript
// Wallet.js vp_token decoded JWT claims

{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://identity.foundation/presentation-exchange/submission/v1"
  ],
  "type": ["VerifiablePresentation", "PresentationSubmission"],
  "presentation_submission": {
    "id": "epzZXstAcVNt5MRrcyG91",
    "definition_id": "IDCardCredentials",
    "descriptor_map": [
      {
        "id": "IDCardCredential",

>         "format": "jwt_vc",

Reviewer 2: this should be `jwt_vp`

>         "path": "$.verifiableCredential[0]"

Reviewer 2: believe this (and below) should be `$.vp.verifiableCredential[0]`

  ]
}
```

4. Implement a `/siopv2/response` endpoint to process and verify the Wallet's response. The HTTP response body must always include a `credential_offer`. If the Wallet's response does not include a `vp_token`, as a valid form of IDV, then the HTTP response body must also include a URL directing the Wallet to an Identity Verification form:

<Shnip
  snippets={[
    {
      snippetName: 'siopv2ResponseEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: 'siopv2ResponseEndpointIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt',
    },
  ]}
/>

:::info

> The IDV Request Form points to a webpage for the customer to input their Personal Identifiable Information(PII) (e.g.; name, identification number, date of birth, etc).  

Reviewer: ```suggestion
The IDV Request Form points to a webpage for the customer to input their Personal Identifiable Information (PII) (e.g.; name, identification number, date of birth, etc).  
```

:::

5. Configure the `/.well-known/openid-credential-issuer` endpoint that will return the Issuer's Metadata:

<Shnip
  snippets={[
    {
      snippetName: 'wellKnownIssuerMetadataEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: 'wellKnownIssuerMetadataEndpointIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt',
    },
  ]}
/>

- `credential_issuer`: The base URL address of the Issuer's service. It is used by the Wallet to retrieve the Issuer's metadata and the Issuer's Authorization Server Metadata.
- `credential_endpoint`: The endpoint the Wallet will use to submit requests to issue credentials. After obtaining an access token, the Wallet sends a credential request to this endpoint.

> - `credential_configurations_supported`: Defines the format `jwt_vc_json` and the credential.

Reviewer: ```suggestion
- `credential_configurations_supported`: Defines the format `jwt_vc_json`, and the credential, including informative display markup.
```

Content-Type: application/json
{
   "proof": {
      "proof_type": "jwt",
      "jwt": "eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDpqd2s6ZXlKcmRIa2lPaUpQUzFBaUxDSmpjbllpT2lKRlpESTFOVEU1SWl3aWVDSTZJalp0TFU1NE9HWndSMWxUT0ZoV04xZ3hVVmN4YWxKdVdHNHRUV1JxVWpJM1ptSktZVGhGVmt0bmEyTWlmUSMwIiwidHlwIjoib3BlbmlkNHZjaS1wcm9vZitqd3QifQ.eyJhdWQiOiJkaWQ6andrOmV5SnJkSGtpT2lKUFMxQWlMQ0pqY25ZaU9pSkZaREkxTlRFNUlpd2llQ0k2SW5SMGNFUTVhekpzUkhkMVdraG5kMlZZVWw5blVHeFNkR1pPYTBGVGVYWXpNR1ZYUkhseVVFOVJNVVVpZlEiLCJpYXQiOjE3MTA4NjMxOTEsIm5vbmNlIjoidFppZ25zbkZicCJ9.MqxVdgX2tJMDJhNYuniYE7bjLyybRgRcPlDCY2ZN9FU2acLC5kU_wAZq-IF8vfIRXfmhbiFoBbfoPUZNBgdDAw"
   }
}
```

The `proof` JWT contains various claims, most notably a `nonce` to prevent replay attacks. For a detailed look at each field please refer to the [OpenID spec](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-7.2.1.1-2.1.1).

> - `czZCaGRSa3F0MzpnWDFmQmF0M2JW`: The `access_token` that the Wallet recieved from the `/token` endpoint.

Reviewer: ```suggestion
- `Authorization: Bearer {jwt}`: The `access_token` that the Wallet recieved from the `/token` endpoint.
```

---
sidebar_position: 1
title: Issuer Flow
hide_title: true
---

<LanguageSwitcher languages="JavaScript, Kotlin" />

# Issuing a Known Customer Credential

Known Customer Credentials (KCCs) are [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) designed to streamline the Know Your Customer (KYC) process for tbDEX protocol users. KCCs help in gaining access to PFIs providing regulated financial services. 

For a comprehensive exploration of the practical applications and compliance considerations of KCCs, refer to our [KCC Compliance Guide](/) which provides detailed examples and insights into how KCCs align with certain regulatory requirements, offering valuable context for those overseeing the implementation and management of KCCs.

:::danger Important
This guide is intended for educational purposes only and does not constitute legal advice. Compliance programs may have varying requirements. Consult your legal and/or compliance advisors to ensure that the KCC is consistent with your legal and compliance obligations.
:::

In this guide, we'll cover:

> - Performing Identity Verification (IDV) as a necessary step **before** issuing a Known Customer Credential (KCC), following [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) and [OpenID for Verifiable Presentation's (OID4VP)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html) set of specifications.

Reviewer: bit of a nit, but the word "following" is used here and two lines down in two different ways (this one means "compatible" and the next one means "subsequently"), maybe this line could be reworded to...

```suggestion
- Performing Identity Verification (IDV) as a necessary step **before** issuing a Known Customer Credential (KCC), by implementing an API compatible with [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) and [OpenID for Verifiable Presentation's (OID4VP)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html).
```

@decentralgabe what is the proper speak in the specification world for "compatible with?" Or is that proper?

Reviewer 2: usually we use the word _conformance_ so something like _conformant with_ would work

- Designing a KCC, detailing the required properties.

> - Issuing a KCC, following successful IDV.

Reviewer: Could we reference and link to OID4VCI in this bullet point (as we did for the other two specs in the first bullet point)?

https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html

Author: @KendallWeihe for this one you're saying we should link to it here?: 

Issuing a KCC, following successful [IDV](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html).

Reviewer: ```suggestion
- Performing Identity Verification (IDV) as a necessary step **before** issuing a Known Customer Credential (KCC), following [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) and [OpenID for Verifiable Presentation's (OID4VP)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html) set of specifications.
- Designing a KCC, detailing the required properties.
- Issuing a KCC, conformant with [OID4VCI](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html), following successful IDV.
```

:::info
For a detailed guide on managing Known Customer Credentials as a Wallet application, please refer to the [Known Customer Credential Guide for Wallet Apps](/)
:::

## Environment Setup

If you haven't already, please follow the [Credential Issuance Server Setup Guide](/) for detailed instructions on the dependencies and packages needed to set up your server.

## IDV Process

Identity Verification (IDV) is a critical component of typical Know Your Customer (KYC) requirements, where certain types of Personally Identifying Information (PII) are collected and verified from an individual. IDV is a crucial step before a Known Customer Credential (KCC) can be issued.

[**IDV Flow diagram**]

> The IDV flow begins when the customer's agent (e.g. Wallet application) sends an HTTP request to the Issuer's [IDV service endpoint](https://hackmd.io/4cW3D9BVQGCEuf1CEg47jA?both#Issuers-Identity) specified in the Issuer's [Decentralized Identifier (DID) Document](/docs/glossary#did-document).

Reviewer 2: should this point to something besides hackmd?

Author: Yes all the links will be updated before we merge next week.

### Handling Incoming Requests

> 1. Implement an endpoint to handle incoming GET requests from the Wallet. When a request is received, you'll need to construct the [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) Authorization Request.

Reviewer: ```suggestion
1. Implement an endpoint to handle incoming GET requests from the Wallet. When a request is received, you'll need to construct the [SIOPv2 Authorization Request](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-9).
```

### Approach 1: Requesting only [`id_token`](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-self-issued-id-token)

> This approach is used when the goal is solely to authenticate the user without requesting any additional Verifiable Credentials:

Reviewer: ```suggestion
This approach is used when a Verifiable Presentation is not an accepted form of IDV:
```

<Shnip
  snippets={[
    {
      snippetName: 'siopv2InitiateId_tokenEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js for id_token only ',
    },
    {
      snippetName: 'siopv2InitiateId_tokenEndpointIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt for id_token only',
    },
  ]}
/>

:::note

> `id_token` is a Self Issued Identity Token, that authenticates the customer to the Issuer by sharing the customer's Decentralized Identifier. This guide will cover an example `id_token` to learn more refer to [openID's specs](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-self-issued-id-token).

Reviewer 2: ```suggestion
`id_token` is a Self Issued Identity Token, that authenticates the customer to the Issuer by sharing the customer's Decentralized Identifier. This guide will cover an example `id_token` to learn more refer to [OpenID's specs](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-self-issued-id-token).
```

:::

### Approach 2: Requesting both `id_token` and [`vp_token`](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2)

> This approach is used when you need to authenticate the user and also want to request specific Verifiable Credentials:

Reviewer: ```suggestion
This approach is used when a Verifiable Presentation **is** an accepted form of IDV:
```

- **`response_type`**: Lets the Wallet know the desired format, with `id_token` being required identity token and `vp_token` for _optional_ verifiable presentation token. The inclusion of `vp_token` means a presentation definition is required within the request.
- **`response_uri`**: Indicates the callback url for the Wallet, ensuring direct communication between the Wallet and the Issuer.
- **`nonce`**: A unique identifier tying together the request and its response.

:::note
`vp_token` is a [Verifiable Presentation](/docs/glossary#verifiable-presentation) that contains certified and authenticated claims about the customer. In this guide, we'll cover an example `vp_token`. To learn more refer to [OpenID's specs](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2).
:::

2. Encode the SIOPv2 Authorization Request and respond:

<Shnip
  snippets={[
    {
      snippetName: 'encodeSiopv2AuthRequestIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: 'encodeSiopv2AuthRequestIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt',
    },
  ]}
/>

> 3. The Wallet will then respond with a POST SIOPv2 Authorization Response. This response will be [JSON Web Tokens (JWTs)](https://jwt.io/). Here's an example response where `vp_token` and `id_token` is in the Authorization Response:

Reviewer: ```suggestion
3. The Wallet will then respond with a POST [SIOPv2 Authorization Response](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-10). The HTTP request body must include an `id_token` and optionally a `vp_token`, both of which must be [JSON Web Tokens (JWTs)](https://datatracker.ietf.org/doc/html/rfc7519). Here are examples of `id_token` and `vp_token`:
```

Reviewer: What do you think about this instead...? (my concern is the reader may think the http request body is of type `text/plain` with the encoded JWT, or that decoded examples include the JWS Headers & Signature)

---

3. The Wallet will then respond with an HTTP POST [SIOPv2 Authorization Response](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-10). The HTTP request body must include an `id_token` and optionally a `vp_token`, both of which must be [JSON Web Tokens (JWTs)](https://datatracker.ietf.org/doc/html/rfc7519). Here is an example HTTP POST request body (which includes the optional `vp_token`):

```javascript
// Wallet.js example HTTP POST request body

{
    "id_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJkaWQ6ZGh0OmN1c3RvbWVyRGlkIiwic3ViIjoiZGlkOmRodDpjdXN0b21lckRpZCIsImF1ZCI6Imlzc3VlckRpZC51cmkiLCJub25jZSI6Im4tMFM2X1d6QTJNaiIsImV4cCI6MTYxODg4NDQ3MywiaWF0IjoxNjE4ODgwODczfQGUuY29tL2NyZWRlbnRpYWxzLzE4NzIiLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiSURDYXJkQ3JlZGVudGlhbCJdLCJpc3N1ZXIiOnsiaWQiOiJkaWQ6ZXhhbXBsZTppc3N1ZXIifSwiaXNzdWFuY2VEYXRlIjoiMjAxMC0wMS0wMVQxOToyMzoyNFoiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJnaXZlbl9uYW1lIjoiRnJlZHJpayIsImZhbWlseV9uYW1.Imlzc3VlckRpZC51cmkiLCJub25jZSI6Im4tMFM2X1d6QTJNaiIsImV4cCI6MTYxODg4NDQ3MywiaWF0IjoxNjE4ODgwODcz",
    "vp_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSJdLCJ0eXBlIjpbIlZlcmlmaWFibGVQcmVzZW50YXRpb24iXSwidmVyaWZpYWJsZUNyZWRlbnRpYWwiOlt7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIiwiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvZXhhbXBsZXMvdjEiXSwiaWQiOiJodHRwczovL2V4YW1wbGUuY29tL2NyZWRlbnRpYWxzLzE4NzIiLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiSURDYXJkQ3JlZGVudGlhbCJdLCJpc3N1ZXIiOnsiaWQiOiJkaWQ6ZXhhbXBsZTppc3N1ZXIifSwiaXNzdWFuY2VEYXRlIjoiMjAxMC0wMS0wMVQxOToyMzoyNFoiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJnaXZlbl9uYW1lIjoiRnJlZHJpayIsImZhbWlseV9uYW1lIjoiU3RyJm9tYmVyZyIs.ImJpcnRoZGF0ZSI6IjE5NDktMDEtMjIifX1dLCJob2xkZXIiOiJkaWQ6ZXhhbXBsZTpob2xkZXIiLCJwcm9vZiI6eyJ0eXBlIjoiRWQyNTUxOVNpZ25hdHVyZTIwMTgiLCJjcmVhdGVkIjoiMjAyMS0wMy0xOVQxNTozMDoxNVoiLCJjaGFsbGVuZ2UiOiJuLTBTNl9XekEyTWoiLCJkb21haW4iOiJodHRwczovL2NsaWVudC5leGFtcGxlLm9yZy9jYiIsImp3cyI6ImV5SmhiR2NpT2lKSVV6STFOaUo5LkhvbGRlcl9GT1JfU0lHTkFUVVJFIiwicHJvb2ZQdXJwb3NlIjoiYXV0aGVudGljYXRpb24iLCJ2ZXJpZmljYXRpb25NZXRob2QiOiJkaWQ6ZXhhbXBsZTpob2xkZXIja2V5LTEifX0"
}
```

The decoded [JWT Claims](https://datatracker.ietf.org/doc/html/rfc7519#section-4) for the `id_token` and `vp_token` are as follows:

```javascript
// Wallet.js id_token decoded JWT claims

{
    "iss": "did:dht:customerDid",
    "sub": "did:dht:customerDid", // Customer's DID string
    "aud": "did:dht:issuerDid", // Issuer's DID string
    "nonce": "n-0S6_WzA2Mj", // Unique string from the Issuer's Auth request
    "exp": 1618884473, // Expiration time
    "iat": 1618880873 // issued at time
}
```

```javascript
// Wallet.js vp_token decoded JWT claims

{
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://identity.foundation/presentation-exchange/submission/v1"
    ],
    "type": ["VerifiablePresentation", "PresentationSubmission"],
    "presentation_submission": {
      "id": "epzZXstAcVNt5MRrcyG91",
      "definition_id": "IDCardCredentials",
      "descriptor_map": [
        {
          "id": "IDCardCredential",
          "format": "jwt_vc",
          "path": "$.verifiableCredential[0]"
        },
        {
          "id": "nationalIdentifierVerification",
          "format": "jwt_vc",
          "path": "$.verifiableCredential[1]"
        }
      ]
    },
    "verifiableCredential": [
      "eyJhbGciOiJFZERTQSIsInR5cCI6IkpXVCIsImtpZCI6ImRpZDprZXk6ejZNa2VyNDlDbnVnN2hzdkhEZ3Y0NHl2cGR2dE1oNHlMaURYeFM2N2huclVodHQyI3o2TWtlcjQ5Q251Zzdoc3ZIRGd2NDR5dnBkdnRNaDR5TGlEWHhTNjdobnJVaHR0MiJ9.eyJpc3MiOiJkaWQ6a2V5Ono2TWtlcjQ5Q251Zzdoc3ZIRGd2NDR5dnBkdnRNaDR5TGlEWHhTNjdobnJVaHR0MiIsInZjIjp7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIl0sInR5cGUiOlsiVmVyaWZpYWJsZUNyZWRlbnRpYWwiLCJFbXBsb3ltZW50Q3JlZGVudGlhbCJdLCJpZCI6InVybjp1dWlkOjcyNDhiOTkyLTkwOTYtNDk2NS1hMGVjLTc3ZDhhODNhMWRmYiIsImlzc3VlciI6ImRpZDprZXk6ejZNa2VyNDlDbnVnN2hzdkhEZ3Y0NHl2cGR2dE1oNHlMaURYeFM2N2huclVodHQyIiwiaXNzdWFuY2VEYXRlIjoiMjAyMy0xMi0yMVQyMDoxMToyNVoiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJpZCI6ImRpZDppb246RWlEMTR4UmY0cTJNWlh1ZWY2X2ZXYnBGbVlTUG94dGFxTkp1SmdEMG96Wl84UTpleUprWld4MFlTSTZleUp3WVhSamFHVnpJanBiZXlKaFkzUnBiMjRpT2lKeVpYQnNZV05sSWl3aVpHOWpkVzFsYm5RaU9uc2ljSFZpYkdsalMyVjVjeUk2VzNzaWFXUWlPaUprZDI0dGMybG5JaXdpY0hWaWJHbGpTMlY1U25kcklqcDdJbU55ZGlJNklrVmtNalUxTVRraUxDSnJkSGtpT2lKUFMxQWlMQ0o0SWpvaWVubGFNbVYzTlhKeVVXdFVjbUV3WlZsVk16WlBTblJzTURCbFJWZHhhalZhV0dkNmNEZFpSVTVKUVNKOUxDSndkWEp3YjNObGN5STZXeUpoZFhSb1pXNTBhV05oZEdsdmJpSmRMQ0owZVhCbElqb2lTbk52YmxkbFlrdGxlVEl3TWpBaWZTeDdJbWxrSWpvaVpIZHVMV1Z1WXlJc0luQjFZbXhwWTB0bGVVcDNheUk2ZXlKamNuWWlPaUp6WldOd01qVTJhekVpTENKcmRIa2lPaUpGUXlJc0luZ2lPaUpQZDJZMFQyMUViamxKWm5SNFdYWnBkRTFHWm1jMVVXeDVMVVV6VWs1b1dsUkdPVlpFTWtnNVQzVjNJaXdpZVNJNkltUnZjVmxtV2s1c1NtRlRNVll4U201bU9HdEZObEF6VkRsd2QzaDNla3hFVTJWc1ZqTlRUa2s1U2xFaWZTd2ljSFZ5Y0c5elpYTWlPbHNpYTJWNVFXZHlaV1Z0Wlc1MElsMHNJblI1Y0dVaU9pSktjMjl1VjJWaVMyVjVNakF5TUNKOVhTd2ljMlZ5ZG1salpYTWlPbHQ3SW1sa0lqb2laSGR1SWl3aWMyVnlkbWxqWlVWdVpIQnZhVzUwSWpwN0ltVnVZM0o1Y0hScGIyNUxaWGx6SWpwYklpTmtkMjR0Wlc1aklsMHNJbTV2WkdWeklqcGJJbWgwZEhCek9pOHZaSGR1TG5SaVpHUmxkaTV2Y21jdlpIZHVOaUlzSW1oMGRIQnpPaTh2WkhkdUxuUmlaR1JsZGk1dmNtY3ZaSGR1TUNKZExDSnphV2R1YVc1blMyVjVjeUk2V3lJalpIZHVMWE5wWnlKZGZTd2lkSGx3WlNJNklrUmxZMlZ1ZEhKaGJHbDZaV1JYWldKT2IyUmxJbjFkZlgxZExDSjFjR1JoZEdWRGIyMXRhWFJ0Wlc1MElqb2lSV2xEWm05bVFUQkpVbU5uY2tWdVVHZHdRbU5RV1ZsV2VFWlliR0pTYjJRd2RVNWZRVkJwTkVrNUxVRmZRU0o5TENKemRXWm1hWGhFWVhSaElqcDdJbVJsYkhSaFNHRnphQ0k2SWtWcFFtd3pWWG80VldGT2REZGxlREJKYjJJMFJFNXNhbFJGVmpaelQwTmtjbFJ3TWxvNE5FTkJPVFJPUWtFaUxDSnlaV052ZG1WeWVVTnZiVzFwZEcxbGJuUWlPaUpGYVVOWk9WRldZbWRKYkUxemRraEZYMVJtTld4a1MxQjBkR3d3WVV4blNrdHNSbmt6Vms0d2QzQTJhVFpSSW4xOSIsImVtcGxveW1lbnRTdGF0dXMiOiJlbXBsb3llZCJ9fX0.Sazc8Ndhs-NKjxvtVMKeC9dxjEkI26fVsp2kFNWM-SYLtxMzKvl5ffeWd81ysHgPmBBSk2ar4dMqGgUsyM4gAQ",
      "eyJhbGciOiJFZERTQSIsInR5cCI6IkpXVCIsImtpZCI6ImRpZDprZXk6ejZNa2pwUzRHVUFoYmdCSmg2azJnZTZvWTQ0UUxyRXA3NXJadHNqYVRLb3JSRGR0I3o2TWtqcFM0R1VBaGJnQkpoNmsyZ2U2b1k0NFFMckVwNzVyWnRzamFUS29yUkRkdCJ9.eyJpc3MiOiJkaWQ6a2V5Ono2TWtqcFM0R1VBaGJnQkpoNmsyZ2U2b1k0NFFMckVwNzVyWnRzamFUS29yUkRkdCIsInZjIjp7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIl0sInR5cGUiOlsiVmVyaWZpYWJsZUNyZWRlbnRpYWwiLCJOYW1lQW5kRG9iQ3JlZGVudGlhbCJdLCJpZCI6InVybjp1dWlkOjliZjM2YzY5LTI0ODAtNDllZC1iMTYyLTRlZDEwOWE3MTc3NyIsImlzc3VlciI6ImRpZDprZXk6ejZNa2pwUzRHVUFoYmdCSmg2azJnZTZvWTQ0UUxyRXA3NXJadHNqYVRLb3JSRGR0IiwiaXNzdWFuY2VEYXRlIjoiMjAyMy0xMi0yMVQyMDowNjowMVoiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJpZCI6ImRpZDppb246RWlDS2o2M0FyZlBGcEpsb2lTd3gxQUhxVWtpWlNoSDZGdnZoSzRaTl9fZDFtQTpleUprWld4MFlTSTZleUp3WVhSamFHVnpJanBiZXlKaFkzUnBiMjRpT2lKeVpYQnNZV05sSWl3aVpHOWpkVzFsYm5RaU9uc2ljSFZpYkdsalMyVjVjeUk2VzNzaWFXUWlPaUprZDI0dGMybG5JaXdpY0hWaWJHbGpTMlY1U25kcklqcDdJbU55ZGlJNklrVmtNalUxTVRraUxDSnJkSGtpT2lKUFMxQWlMQ0o0SWpvaWNscFdXbTVJVkVrNWFEWkJUVmxVV0dwT01HcFhTVkYwTTI5ak4xTnJTeTF4Y2kxcVVuSTBUalEzUlNKOUxDSndkWEp3YjNObGN5STZXeUpoZFhSb1pXNTBhV05oZEdsdmJpSmRMQ0owZVhCbElqb2lTbk52YmxkbFlrdGxlVEl3TWpBaWZTeDdJbWxrSWpvaVpIZHVMV1Z1WXlJc0luQjFZbXhwWTB0bGVVcDNheUk2ZXlKamNuWWlPaUp6WldOd01qVTJhekVpTENKcmRIa2lPaUpGUXlJc0luZ2lPaUpaVDFwRE5WSmlUMHQ1T0dadVVUWTJVWEZPUkc5aldFMXZPVXhUZEdNNVYyOWthMHd0ZFZCZlExQnZJaXdpZVNJNklsWnZZM0UxVERodFozQlhXVTFrYjFwS1JrWlJUa1ZDT0hsR0xXTndkRWQzZFdkcFRWVm5hR2t6Y21jaWZTd2ljSFZ5Y0c5elpYTWlPbHNpYTJWNVFXZHlaV1Z0Wlc1MElsMHNJblI1Y0dVaU9pSktjMjl1VjJWaVMyVjVNakF5TUNKOVhTd2ljMlZ5ZG1salpYTWlPbHQ3SW1sa0lqb2laSGR1SWl3aWMyVnlkbWxqWlVWdVpIQnZhVzUwSWpwN0ltVnVZM0o1Y0hScGIyNUxaWGx6SWpwYklpTmtkMjR0Wlc1aklsMHNJbTV2WkdWeklqcGJJbWgwZEhCek9pOHZaSGR1TG5SaVpHUmxkaTV2Y21jdlpIZHVOaUlzSW1oMGRIQnpPaTh2WkhkdUxuUmlaR1JsZGk1dmNtY3ZaSGR1TUNKZExDSnphV2R1YVc1blMyVjVjeUk2V3lJalpIZHVMWE5wWnlKZGZTd2lkSGx3WlNJNklrUmxZMlZ1ZEhKaGJHbDZaV1JYWldKT2IyUmxJbjFkZlgxZExDSjFjR1JoZEdWRGIyMXRhWFJ0Wlc1MElqb2lSV2xCTXpSMlMzb3llVmswZVV4dGRDMUdabkJuYWpWbGFFRm1ZWFI1YzFOa2MwNVNWbVpMYkhwUWRqTjVkeUo5TENKemRXWm1hWGhFWVhSaElqcDdJbVJsYkhSaFNHRnphQ0k2SWtWcFF6ZGZjMXBzTW1wMVVXNUdhRVJIV1RSb2NFVTRiMlF4YVU5MWRuZG1PVFJ5TVVkbk9HMWFWbVJCVmxFaUxDSnlaV052ZG1WeWVVTnZiVzFwZEcxbGJuUWlPaUpGYVVKdU5sTnJiSEpWYzNKdVFuaFJPVXBqVXkxTlNVaGtWelYwTXpRM1MxWjNaMXBwVEZwMFQwcDRRVkYzSW4xOSIsIm5hbWUiOiJhbGljZSBib2IiLCJkYXRlT2ZCaXJ0aCI6IjEwLTAxLTE5OTAifX19.mNCDv_JntH-wZpYONKNL58UbOWaYXCYJO_HPI_WVlSgwzo6dhYmV_9qtpFKd_exFb-aaEYPeSE43twWlrJeSBg"
    ]
}
```

Reviewer 2: I like pulling it out like this

```javascript
// Wallet.js id_token JWT response

eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9
  .eyJpc3MiOiJkaWQ6ZGh0OmN1c3RvbWVyRGlkIiwic3ViIjoiZGlkOmRodDpjdXN0b21lckRpZCIsImF1ZCI6Imlzc3VlckRpZC51cmkiLCJub25jZSI6Im4tMFM2X1d6QTJNaiIsImV4cCI6MTYxODg4NDQ3MywiaWF0IjoxNjE4ODgwODczfQGUuY29tL2NyZWRlbnRpYWxzLzE4NzIiLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiSURDYXJkQ3JlZGVudGlhbCJdLCJpc3N1ZXIiOnsiaWQiOiJkaWQ6ZXhhbXBsZTppc3N1ZXIifSwiaXNzdWFuY2VEYXRlIjoiMjAxMC0wMS0wMVQxOToyMzoyNFoiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJnaXZlbl9uYW1lIjoiRnJlZHJpayIsImZhbWlseV9uYW1

>   .Imlzc3VlckRpZC51cmkiLCJub25jZSI6Im4tMFM2X1d6QTJNaiIsImV4cCI6MTYxODg4NDQ3MywiaWF0IjoxNjE4ODgwODcz;

Reviewer: I think this should maybe all be on one line? not sure how the markup works here

Reviewer: Oh okay, I see how the markup renders this, yeah this is good. Disregard.

<img width="699" alt="Screenshot 2024-03-17 at 2 27 33 PM" src="https://github.com/TBD54566975/developer.tbd.website/assets/3602993/e4e73349-4d30-4360-a53a-ec4c3a93b8b1">

    ]
  }
}
```

> 4. Implement a `/siopv2/response` endpoint to process and verify the Wallet's response. If the Wallet's response includes a [Verifiable Presentation](/docs/glossary#verifiable-presentation) (`vp_token`), you, as the Issuer, will respond with a `credential_offer`. If the Wallet's response only includes an `id_token`, the `credential_offer` will also include a URL directing to an Identity Verification form:

Reviewer: ```suggestion
4. Implement a `/siopv2/response` endpoint to process and verify the Wallet's response. The HTTP response body must always include a `credential_offer`. If the Wallet's response does not include a `vp_token`, as a valid form of IDV, then the HTTP response body must also include a URL directing the Wallet to an Identity Verification form:
```

Some of this language is duplicative of the `:::info` note you have below, so feel free to wrangle it how you think best, but just want to call out that URL is not within the credential offer but beside it ([more info here](https://github.com/TBD54566975/known-customer-credential#idv-request))

<Shnip
  snippets={[
    {
      snippetName: 'siopv2ResponseEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: 'siopv2ResponseEndpointIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt',
    },
  ]}
/>

:::info
When the Wallet's response solely contains `id_token`, your response must include a **`url`** (IDV Request Form). This will point to a webpage for the customer to input their information (e.g.; name, identification number, date of birth, etc).  
:::

#### Credential Offer Details:

- **`credential_issuer`**: The base URL address of the Issuer's service. It is used by the Wallet to retrieve the Issuer's metadata and the Issuer's Authorization Server Metadata.
- **`credential_configuration_ids`**: Unique strings that identify the credential you are offering. The Wallet can use these to request metadata.
- **`grants`**: Contains grant types you will accept for this credential offer.

> - **`urn:ietf:params:oauth:grant-type:pre-authorized_code`**: Grant type that allows the Wallet to follow a Pre-Authorized code flow skipping any PII steps because they've already completed it.

Reviewer: ```suggestion
- **`urn:ietf:params:oauth:grant-type:pre-authorized_code`**: Grant type required for the [OID4VCI](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#appendix-B.1) Pre-Authorization flow.
```

- **`pre-authorized_code`**: The code representing the Credential Issuer's authorization for the Wallet to obtain an access token.

:::info
**Dynamic Metadata Retrieval by Wallets**

> Wallets use the `credential_issuer` from the Credential Offer to dynamically construct URLs for fetching Issuer and Authorization Server Metadata. These URLs, formed by appending `/.well-known/openid-credential-issuer` and` /.well-known/oauth-authorization-server` to the `credential_issuer`, allow wallets to access up-to-date information on issuer capabilities and OAuth 2.0 endpoints. The next steps will cover how to properly configure these endpoints.

Reviewer: lil space 

```suggestion
Wallets use the `credential_issuer` from the Credential Offer to dynamically construct URLs for fetching Issuer and Authorization Server Metadata. These URLs, formed by appending `/.well-known/openid-credential-issuer` and `/.well-known/oauth-authorization-server` to the `credential_issuer`, allow wallets to access up-to-date information on issuer capabilities and OAuth 2.0 endpoints. The next steps will cover how to properly configure these endpoints.
```

For a comprehensive exploration of the practical applications and compliance considerations of KCCs, refer to our [KCC Compliance Guide](/) which provides detailed examples and insights into how KCCs align with certain regulatory requirements, offering valuable context for those overseeing the implementation and management of KCCs.

:::danger Important
This guide is intended for educational purposes only and does not constitute legal advice. Compliance programs may have varying requirements. Consult your legal and/or compliance advisors to ensure that the KCC is consistent with your legal and compliance obligations.
:::

In this guide, we'll cover:

- Performing Identity Verification (IDV) as a necessary step **before** issuing a Known Customer Credential (KCC), by implementing an HTTP API conformant with [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) and [OpenID for Verifiable Presentation's (OID4VP)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html).

> - Designing a KCC, detailing the required properties.

Reviewer 3: Designing and detailing the required properties of a KCC

> - Issuing a KCC, conformant with [OpenID for Verifiable Credential Issuance (OID4VCI)](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html), following successful IDV.

Reviewer 3: nit: remove commas

> 

Reviewer 3: Since this is an issuer-focused guide, I think we may want to add something here to the effect of "by the time you've completed this guide, you'll have the technical tool to perform IDV as well as design and issue a KCC."

> :::info

Reviewer 3: Do we need to take this out pending tbdex.io launch?

Author: not sure what you mean?

Reviewer 3: There's no destination link, and I imagine that link would be to tbdex.io, so I was wondering if we need to wait to include this note until that endpoint is available

Author: ah ok I thought everything was being pushed and going live at the same time, and was going to update the link before pushing. Did we plan on putting this out before the website is live?

For a detailed guide on managing Known Customer Credentials as a Wallet application, please refer to the [Known Customer Credential Guide for Wallet Apps](/)
:::

## Environment Setup

> If you haven't already, please follow the [Credential Issuance Server Setup Guide](/) for detailed instructions on the dependencies and packages needed to set up your server.

Reviewer 3: Missing link?

Author: thanks, adding all missing links before we push this out

## IDV Process

> Identity Verification (IDV) is a critical component of typical Know Your Customer (KYC) requirements, where certain types of Personally Identifying Information (PII) are collected and verified from an individual. IDV is a crucial step before a Known Customer Credential (KCC) can be issued.

Reviewer 3: I don't think we need to re-state abbreviations here since we already used them earlier in the article

> [**IDV Flow diagram**]

Reviewer 3: I will add this

The IDV flow begins when the customer's agent (e.g. Wallet application) sends an HTTP request to the Issuer's [IDV service endpoint](https://hackmd.io/4cW3D9BVQGCEuf1CEg47jA?both#Issuers-Identity) specified in the Issuer's [Decentralized Identifier (DID) Document](/docs/glossary#did-document).

### Handling Incoming Requests

> 1. Implement an endpoint to handle incoming GET requests from the Wallet. When a request is received, you'll need to construct the [SIOPv2 Authorization Request](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-9).

Reviewer 3: Remove numbering?

### Approach 1: Requesting only [`id_token`](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-self-issued-id-token)

> This approach is used when a Verifiable Presentation is **not** an accepted form of IDV:

Reviewer 3: Link or explainer on what this is

> This approach is used when a Verifiable Presentation is **not** an accepted form of IDV:

Reviewer 3: When a Verifiable Presentation is not an accepted form of IDV, you can respond to the customer's Wallet application by requesting only an `id_token` from them:

<Shnip
  snippets={[
    {
      snippetName: 'siopv2InitiateId_tokenEndpointIssuerJs',
      language: 'JavaScript',

>       title: 'Issuer api.js for id_token only ',

Reviewer 3: Do we need separate titles for each of these, or can we have one title above this?

    },
    {
      snippetName: 'siopv2InitiateId_tokenEndpointIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt for id_token only',
    },
  ]}
/>

:::note
`id_token` is a Self Issued Identity Token, that authenticates the customer to the Issuer by sharing the customer's Decentralized Identifier. This guide will cover an example `id_token` to learn more refer to [OpenID's specs](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-self-issued-id-token).
:::

### Approach 2: Requesting both `id_token` and [`vp_token`](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2)

> This approach is used when a Verifiable Presentation **is** an accepted form of IDV:

Reviewer 3: When a Verifiable Presentation is an accepted form of IDV, you can respond to the customer's Wallet application by requesting an `id_token` and a `vp_token` from them:

<Shnip
  snippets={[
    {
      snippetName: 'siopv2InitiateId_tokenAndVp_tokenEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js for id_token and vp_token',
    },
    {
      snippetName: 'siopv2InitiateId_tokenAndVp_tokenEndpointIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt for id_token and vp_token',
    },
  ]}
/>

> The SIOPv2 Authorization request contains information about you, the Issuer, along with a [Presentation Definition](/docs/web5/build/verifiable-credentials/presentation-definition).

Reviewer 3: Can you clarify here whether the Authorization request is the request you as the issuer sent or the requests the Wallet sent? The reader might not have internalized the diagram or other info yet so it can be helpful to restate:

"The SOIPv2 Authorization request above that you as the issuer send to your customer contains the following key information about you as well as the Presentation Definition"

- **`client_id`**: Issuer's DID, establishing the issuer/requestor's identity.
- **`response_type`**: Lets the Wallet know the desired format, with `id_token` being required identity token and `vp_token` for _optional_ verifiable presentation token. The inclusion of `vp_token` means a presentation definition is required within the request.
- **`response_uri`**: Indicates the callback url for the Wallet, ensuring direct communication between the Wallet and the Issuer.
- **`nonce`**: A unique identifier tying together the request and its response.

:::note
`vp_token` is a [Verifiable Presentation](/docs/glossary#verifiable-presentation) that contains certified and authenticated claims about the customer. In this guide, we'll cover an example `vp_token`. To learn more refer to [OpenID's specs](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2).
:::

> 2. Encode the SIOPv2 Authorization Request and respond:

Reviewer 3: The numbering is a little confusing since it seems like the above is really 1a) or 1b) - can we clarify the numbering or nix the numbering in favor of using headings?

> 2. Encode the SIOPv2 Authorization Request and respond:

Reviewer 3: Can we do a `...` or something similar to compress the whole `SOIPv2` object above so it isn't taking up so much space? That way we can focus on what's new

Author: it won't compile that way, that's why I think I highlighted it

    "vp_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSJdLCJ0eXBlIjpbIlZlcmlmaWFibGVQcmVzZW50YXRpb24iXSwidmVyaWZpYWJsZUNyZWRlbnRpYWwiOlt7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIiwiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvZXhhbXBsZXMvdjEiXSwiaWQiOiJodHRwczovL2V4YW1wbGUuY29tL2NyZWRlbnRpYWxzLzE4NzIiLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiSURDYXJkQ3JlZGVudGlhbCJdLCJpc3N1ZXIiOnsiaWQiOiJkaWQ6ZXhhbXBsZTppc3N1ZXIifSwiaXNzdWFuY2VEYXRlIjoiMjAxMC0wMS0wMVQxOToyMzoyNFoiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJnaXZlbl9uYW1lIjoiRnJlZHJpayIsImZhbWlseV9uYW1lIjoiU3RyJm9tYmVyZyIs.ImJpcnRoZGF0ZSI6IjE5NDktMDEtMjIifX1dLCJob2xkZXIiOiJkaWQ6ZXhhbXBsZTpob2xkZXIiLCJwcm9vZiI6eyJ0eXBlIjoiRWQyNTUxOVNpZ25hdHVyZTIwMTgiLCJjcmVhdGVkIjoiMjAyMS0wMy0xOVQxNTozMDoxNVoiLCJjaGFsbGVuZ2UiOiJuLTBTNl9XekEyTWoiLCJkb21haW4iOiJodHRwczovL2NsaWVudC5leGFtcGxlLm9yZy9jYiIsImp3cyI6ImV5SmhiR2NpT2lKSVV6STFOaUo5LkhvbGRlcl9GT1JfU0lHTkFUVVJFIiwicHJvb2ZQdXJwb3NlIjoiYXV0aGVudGljYXRpb24iLCJ2ZXJpZmljYXRpb25NZXRob2QiOiJkaWQ6ZXhhbXBsZTpob2xkZXIja2V5LTEifX0"
}
```

> The decoded [JWT Claims](https://datatracker.ietf.org/doc/html/rfc7519#section-4) for the `id_token` and `vp_token` are as follows:

Reviewer 3: I realized here that the only definition of the `id_token` comes in a note, and there is no definition for `vp_token`. Can we add clear definitions or links for both of them at the top of these instructions?

  ]
}

> ```

Reviewer 3: To save vertical space let's put these in breadcrumbs or tabs

> 4. Implement a `/siopv2/response` endpoint to process and verify the Wallet's response. The HTTP response body must always include a `credential_offer`. If the Wallet's response does not include a `vp_token`, as a valid form of IDV, then the HTTP response body must also include a URL directing the Wallet to an Identity Verification form:

Reviewer 3: is this a standard endpoint name, or just something that makes the most sense?

> 4. Implement a `/siopv2/response` endpoint to process and verify the Wallet's response. The HTTP response body must always include a `credential_offer`. If the Wallet's response does not include a `vp_token`, as a valid form of IDV, then the HTTP response body must also include a URL directing the Wallet to an Identity Verification form:

Reviewer 3: remove comma

> <Shnip

Reviewer 3: Is there any way we can break these snippets up, even using breadcrumbs? The comment headers are super helpful, wanted to see if you had any ideas

  snippets={[
    {
      snippetName: 'siopv2ResponseEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: 'siopv2ResponseEndpointIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt',
    },
  ]}
/>

:::info
The IDV Request Form points to a webpage for the customer to input their Personal Identifiable Information (PII) (e.g.; name, identification number, date of birth, etc).  
:::

> #### Credential Offer Details:

Reviewer 3: It's unclear what the credential offer is - can we include a definition here or a reference prior to this section since it is referred to in the code?

- **`credential_configuration_ids`**: Unique strings that identify the credential you are offering. The Wallet can use these to request metadata.
- **`grants`**: Contains grant types you will accept for this credential offer.
- **`urn:ietf:params:oauth:grant-type:pre-authorized_code`**: Grant type required for the [OID4VCI](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#appendix-B.1) Pre-Authorization flow.
- **`pre-authorized_code`**: The code representing the Credential Issuer's authorization for the Wallet to obtain an access token.

:::info
**Dynamic Metadata Retrieval by Wallets**

Wallets use the `credential_issuer` from the Credential Offer to dynamically construct URLs for fetching Issuer and Authorization Server Metadata. These URLs, formed by appending `/.well-known/openid-credential-issuer` and `/.well-known/oauth-authorization-server` to the `credential_issuer`, allow wallets to access up-to-date information on issuer capabilities and OAuth 2.0 endpoints. The next steps will cover how to properly configure these endpoints.
:::

> 5. Configure the `/.well-known/openid-credential-issuer` endpoint that will return the Issuer's Metadata:

Reviewer 3: your Issuer metadata

      language: 'Kotlin',
      title: 'Issuer Application.kt',
    },
  ]}
/>

- `issuer`: The base URL of the Issuer's service.
- `token_endpoint`: The endpoint where Wallet's can exchange the `pre_authorization_code` receieved in the `credential_offer` for an access token.

7. Set up the `/token` endpoint to process incoming requests from the Wallet containing the `pre_authorization_code`. This code, originally provided in the `credential_offer` object, is used to validate the Wallet's request:

<Shnip
  snippets={[
    {
      snippetName: 'accessTokenEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: 'accessTokenEndpointIssuerKT',
      language: 'Kotlin',
      title: 'Issuer Application.kt',
    },
  ]}
/>

:::info

**`authorization_pending` error response in the case IDV is incomplete**

The Issuer must check if the IDV Process is completed prior to issuing an access token, and in the event IDV is incomplete must return an `authorization_pending` error response to the Wallet whereafter the Wallet will repeat the call to `/token` until IDV is completed.
:::

## Design Known Customer Credential

> When creating a Verifiable Credential, you can design a model class to represent the specific type of credential you'd like to issue.

Reviewer 3: Link to Web5 docs with background info on credentials and credential design

Reviewer 3: We also want to tell devs what does and doesn't go into this design!

Author: What do you mean? what doesn't go into it that we need to call out? I think showing the class and the `info` tells them what goes in it? 

Also yea I can link to the VC learn guide again here, what guide do we have that goes over design?

Reviewer 3: To clarify, I think it may be helpful to have an example of what may go into this in the plaintext here in addition to the class. That way the text and code stay connected

1. Once the customer accepts your Credential Offer, the Wallet will then send a request for credentials to the endpoint specified in the `credential_endpoint` field of the Issuer's metadata. Here is an example of the Wallet's request:

```bash
POST /credentials HTTP/1.1
Host: issuer.example.com
Authorization: Bearer eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDpqd2s6ZXlKcmRIa2lPaUpQUzFBaUxDSmpjbllpT2lKRlpESTFOVEU1SWl3aWVDSTZJblIwY0VRNWF6SnNSSGQxV2tobmQyVllVbDluVUd4U2RHWk9hMEZUZVhZek1HVlhSSGx5VUU5Uk1VVWlmUSMwIn0.eyJleHAiOjE3MTA5NDkzODAsImlhdCI6MTcxMDg2Mjk1NywiaXNzIjoiZGlkOmp3azpleUpyZEhraU9pSlBTMUFpTENKamNuWWlPaUpGWkRJMU5URTVJaXdpZUNJNkluUjBjRVE1YXpKc1JIZDFXa2huZDJWWVVsOW5VR3hTZEdaT2EwRlRlWFl6TUdWWFJIbHlVRTlSTVVVaWZRIiwic3ViIjoiZGlkOmp3azpleUpyZEhraU9pSlBTMUFpTENKamNuWWlPaUpGWkRJMU5URTVJaXdpZUNJNklqWnRMVTU0T0dad1IxbFRPRmhXTjFneFVWY3hhbEp1V0c0dFRXUnFVakkzWm1KS1lUaEZWa3RuYTJNaWZRIn0.5_CFzPpvvbxubtXyz7jsmwqaA3SPcebk9OvSWGnORMMfIgh6iWYj5DjXgTst3Mz7ZLjK0AaCF628hwis5hL3AA
Content-Type: application/json
{
   "proof": {
      "proof_type": "jwt",

>       "jwt": "eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDpqd2s6ZXlKcmRIa2lPaUpQUzFBaUxDSmpjbllpT2lKRlpESTFOVEU1SWl3aWVDSTZJalp0TFU1NE9HWndSMWxUT0ZoV04xZ3hVVmN4YWxKdVdHNHRUV1JxVWpJM1ptSktZVGhGVmt0bmEyTWlmUSMwIiwidHlwIjoib3BlbmlkNHZjaS1wcm9vZitqd3QifQ.eyJhdWQiOiJkaWQ6andrOmV5SnJkSGtpT2lKUFMxQWlMQ0pqY25ZaU9pSkZaREkxTlRFNUlpd2llQ0k2SW5SMGNFUTVhekpzUkhkMVdraG5kMlZZVWw5blVHeFNkR1pPYTBGVGVYWXpNR1ZYUkhseVVFOVJNVVVpZlEiLCJpYXQiOjE3MTA4NjMxOTEsIm5vbmNlIjoidFppZ25zbkZicCJ9.MqxVdgX2tJMDJhNYuniYE7bjLyybRgRcPlDCY2ZN9FU2acLC5kU_wAZq-IF8vfIRXfmhbiFoBbfoPUZNBgdDAw"

Reviewer 3: Do we actually need this snippet in this guide, or just the wallet guide?

Author: yes I'd say so it's good for the issuer to know what the response they need to handle will look like
